op-node: Execution Layer Sync

trianglesphere opened the pull request:
**Description**

This passes unsafe payloads directly the EngineController for immediate
insertion when Execution Layer sync is active. This tells the execution
client to sync to that target. Once the EL sync is complete, the last
unsafe payload is marked as safe. This is required when doing snap sync
because the EL does not have the pre-state required to do the engine
consolidation until the sync is complete.

**Tests**

I've manually tested this code & also have added an action test.


**Metadata**

- Fixes https://github.com/ethereum-optimism/client-pod/issues/377
- Fixes https://github.com/ethereum-optimism/optimism/issues/8308

trianglesphere committed:
op-e2e: Actually run TestSystemE2E

ajsutton commented on op-e2e/actions/sync_test.go:
```diff
@@ -164,43 +165,49 @@ func TestUnsafeSync(gt *testing.T) {
+	verifier.ActL2InsertUnsafePayload(seqHead)(t)
+
+	// Expect snap sync to download & execute the entire chain
+	// Verify this by checking that the verifier has the correct value for block 0
+	require.Eventually(t,
+		func() bool {
+			block, err := verifier.eng.L2BlockRefByNumber(t.Ctx(), 1)
```
The comment and the code don't match up here - it's actually checking block 1 matches (which makes more sense given genesis should always match)

ajsutton commented on op-node/rollup/derive/engine_controller.go:
```diff
@@ -258,6 +291,22 @@ func (e *EngineController) TryUpdateEngine(ctx context.Context) error {
+		if errors.Is(err, ethereum.NotFound) {
+			e.syncStatus = syncStatusStartedEL
+			e.log.Info("Starting EL sync")
+			e.elStart = time.Now()
```
Ideally we'd use the clock abstraction here rather than time directly. It is just logging but nice to make progress towards abstracting time everywhere rather than away from it.

trianglesphere replied:
I switched to using the clock abstraction internally (but haven't threaded it in yet)

ajsutton commented on op-program/client/l2/engineapi/l2_engine_api.go:
```diff
@@ -329,7 +336,24 @@ func (ea *L2EngineAPI) forkchoiceUpdated(ctx context.Context, state *eth.Forkcho
+			ea.log.Warn("Must register downloader to be able to snap sync")
+			return STATUS_SYNCING, nil
+		}
+		// If the head hash is unknown (was not given to us in a newPayload request),
+		// we cannot resolve the header, so not much to do. This could be extended in
+		// the future to resolve from the `eth` network, but it's an unexpected case
+		// that should be fixed, not papered over.
```
This comment was copied from geth right? It happens to be wrong but geth implements the engine API wrong by requiring `engine_newPayload` before `engine_forkChoiceUpdated`. But given we're mirroring geth's functionality here it's fine.

op-program will never actually snap sync so it only affects the e2e action tests.

trianglesphere replied:
Yea, this is stripped down from geth. It's probably good that our e2e tests match geth.

I was kinda surprised op-program actually used this code TBH

ajsutton reviewed (approved):
LGTM.

protolambda commented on op-e2e/actions/l2_sequencer_test.go:
```diff
@@ -16,14 +20,30 @@ import (
+		nodeCfg.P2P = p2p.Config{
+			MaxPeers:    100,
+			NoDiscovery: true,
+			ListenAddr:  "0.0.0.0:0",
```
```suggestion
			ListenAddr:  "127.0.0.1:0",
```
To not bind them to external facing IP. Localhost should be enough

protolambda commented on op-node/rollup/driver/state.go:
```diff
@@ -323,6 +355,10 @@ func (s *Driver) eventLoop() {
 			delayedStepReq = nil
 			step()
 		case <-stepReqCh:
+			// Don't start the derivation pipeline until we are done with EL sync
+			if s.engineController.IsEngineSyncing() {
+				continue
+			}
```
If we never receive an unsafe payload via p2p, and never trigger other engine operations through derivation, do we get back any Engine API response that can get us unstuck from a syncing engine state? We may want to repeat the last forkchoice call periodically just for the sync status response, to get unstuck upon sync completion. A simple solution might be to only `continue` if it's been less than X minutes since last derivation call, in the case the engine is syncing.

trianglesphere replied:
We pretty heavily rely on the gossip network to be working in order to work with snap sync. If there's no gossipped blocks, we can't drive it in the first place.

protolambda replied:
I'm thinking of the reverse case, when p2p stops (e.g. sequencer p2p gossip issue, or p2p issue on user end), and the engine is still syncing. Then the op-node is unable to see the engine sync complete, until the op-node restarts.

protolambda reviewed (commented):
Generally looks good, but I am wondering how we exit a syncing engine status when there is no derivation or p2p activity.

trianglesphere committed:
op-e2e: Snap Sync action test

trianglesphere committed:
op-service/node: Add time.Since to Clock & use it

trianglesphere committed:
op-node: Execution Layer Sync

This passes unsafe payloads directly the EngineController for immediate
insertion when Execution Layer sync is active. This tells the execution
client to sync to that target. Once the EL sync is complete, the last
unsafe payload is marked as safe. This is required when doing snap sync
because the EL does not have the pre-state required to do the engine
consolidation until the sync is complete.
